### PR TITLE
Eval: Move heartbeat verification earlier

### PR DESCRIPTION
**Pull request by jannotti**
Adds needed fields to HeartbeatTxnFields so that the HbProof can be checked with other signatures (early, and concurrently).

### CODE REVIEW

File: ledger/apply/heartbeat.go [hunk 1]
```diff
@@ -105,11 +90,16 @@ func Heartbeat(hb transactions.HeartbeatTxnFields, header transactions.Header, b
...
 	}
 	if hdr.Seed != hb.HbSeed {
-		return fmt.Errorf("provided seed %v does not match round %d's seed %v", hb.HbSeed, header.FirstValid, hdr.Seed)
+		return fmt.Errorf("provided seed %v does not match round %d's seed %v",
+			hb.HbSeed, header.FirstValid, hdr.Seed)
 	}
-
-	if !sv.Verify(id, hdr.Seed, hb.HbProof.ToOneTimeSignature()) {
-		return fmt.Errorf("heartbeat failed verification with VoteID %v", sv)
+	if account.VotingData.VoteID != hb.HbVoteID {
```
Comment (algorandskiy) [inline ledger/apply/heartbeat.go:96]: there is no `account.VotingData.VoteID.IsEmpty` check so is possible to trick and call heartbeat on an account without voting keys? I suggest to have both match and empty checks

File: data/transactions/transaction.go [hunk 2]
```diff
@@ -570,6 +570,9 @@ func (tx Transaction) WellFormed(spec SpecialAddresses, proto config.ConsensusPa
 		if !proto.Heartbeat {
 			return fmt.Errorf("heartbeat transaction not supported")
 		}
+		if tx.HbKeyDilution == 0 {
+			return errors.New("tx.HbKeyDilution is zero")
+		}
```
Comment (algorandskiy) [inline data/transactions/transaction.go:575]: idk, maybe it makes sense to reject HB with empty HbSeed/HbVoteID earlier right here?

File: data/transactions/verify/txn.go [hunk 3]
```diff
@@ -310,6 +310,15 @@ func stxnCoreChecks(gi int, groupCtx *GroupContext, batchVerifier crypto.BatchVe
 		return err
 	}
 
+	if s.Txn.Type == protocol.HeartbeatTx {
```
Comment (algorandskiy) [inline data/transactions/verify/txn.go:313]: I guess missing a unit test for successful and failed verification?

File: data/transactions/transaction.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/transaction.go:575]: Agree, I'm adding those to WellFormed.

File: data/transactions/verify/txn.go (hunk 3 above)
Comment (jannotti) [reply data/transactions/verify/txn.go:313]: I'm trying.  It's much harder to write now, assuming I want to actually run it through the whole batch verifications stxnCore stuff.  I'm adding things the txn_test.go.  Will ping you when I push the commits, then please look to confirm you agree.

File: data/transactions/transaction.go (hunk 2 above)
Comment (jannotti) [reply data/transactions/transaction.go:575]: Agreed, I'm adding empty checks.

File: ledger/apply/heartbeat.go (hunk 1 above)
Comment (jannotti) [reply ledger/apply/heartbeat.go:96]: Going into WellFormed.

File: data/basics/userBalance.go [hunk 4]
```diff
@@ -575,16 +575,21 @@ func (u OnlineAccountData) VotingStake() MicroAlgos {
...
 	}
 
 	return proto.DefaultKeyDilution
 }
 
+// KeyDilution returns the key dilution for this account, returning the default
+// key dilution if not explicitly specified.
+func (u OnlineAccountData) KeyDilution(proto config.ConsensusParams) uint64 {
+	return KeyDilution(proto, u.VoteKeyDilution)
+}
```
[the remark this answers is not in the document]
Comment (jannotti) [reply data/basics/userBalance.go]: Added to WellFormed

File: data/transactions/transaction.go (hunk 2 above)
Comment (gmalouf) [reply data/transactions/transaction.go:575]: Reminder for myself; will look for most of the "is this is a free heartbeat, go draconian" checks (the must be empties) to move here.

**Review by gmalouf — COMMENTED** [review]
I'm good pending the checks moving into `WellFormed" (and test cases around them).

File: data/transactions/transaction.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/transaction.go:604]: I didn't go so far as to check that each individual field of `HbProof` is non zero, though I could see the argument to do so.  Thoughts?

Comment (jannotti) [reply data/transactions/transaction.go:604]: Check now.

File: data/transactions/verify/txn.go (hunk 3 above)
Comment (jannotti) [reply data/transactions/verify/txn.go:313]: See txn_test.go
```
// TestTxnHeartbeat makes sure that a heartbeat transaction is valid (and added
// to the cache) only if the normal outer signature is valid AND the inner
// HbProof is valid.
func TestTxnHeartbeat(t *testing.T) {
```
If I understand it correctly, it tests passing and failing.

File: data/transactions/transaction.go (hunk 2 above)
Comment (jannotti) [reply data/transactions/transaction.go:604]: Should I use the `tx.<FieldName>` construction universally? It's common in WellFormed.

File: data/transactions/transaction.go [hunk 5]
```diff
@@ -570,6 +571,34 @@ func (tx Transaction) WellFormed(spec SpecialAddresses, proto config.ConsensusPa
...
 		}
+
+		// If this is a free/cheap heartbeat, it must be very simple.
+		if tx.Fee.Raw < proto.MinTxnFee && tx.Group.IsZero() {
+			kind := "free"
+			if tx.Fee.Raw > 0 {
+				kind = "cheap"
+			}
+
+			if len(tx.Note) > 0 {
```
Comment (gmalouf) [inline data/transactions/transaction.go:582]: NIT: should we say 'free' heartbeat is not allowed to have these things?

File: data/transactions/transaction.go (hunk 2 above)
Comment (gmalouf) [reply data/transactions/transaction.go:604]: I'd be fine with that.

File: data/transactions/transaction.go (hunk 5 above)
Comment (jannotti) [reply data/transactions/transaction.go:582]: We say free if they paid 0, cheap if they pay less than min-fee.

**Review by gmalouf — APPROVED** [review]

**Pull request merged**